src/community/label_propagation.c

```diff
@@ -26,15 +26,410 @@
+                            ) {
+                            /* Nope, we need at least one more iteration */
+                            running = true;
+                        }
+                    }
+                    else {
+                        /* Select randomly from the dominant labels */
+                        k = RNG_INTEGER(0, igraph_vector_int_size(&dominant_labels) - 1);
+                        VECTOR(*membership)[v1] = VECTOR(dominant_labels)[k];
+                    }
+                }
+            }
+
+            /* Clear the nonzero elements in label_counters */
+            num_neis = igraph_vector_int_size(&nonzero_labels);
+            for (j = 0; j < num_neis; j++) {
+                VECTOR(label_counters)[VECTOR(nonzero_labels)[j]] = 0;
+            }
+        }
+
+        /* Alternating between control iterations and label updating iterations */
+        if (!retention)
+                control_iteration = !control_iteration;
+    }
+
+    RNG_END();
+
+    if (weights) {
+        igraph_inclist_destroy(&il);
+    } else {
+        igraph_adjlist_destroy(&al);
+    }
+    IGRAPH_FINALLY_CLEAN(1);
+
+    igraph_vector_int_destroy(&node_order);
+    igraph_vector_int_destroy(&nonzero_labels);
+    igraph_vector_int_destroy(&dominant_labels);
+    igraph_vector_destroy(&label_counters);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+igraph_error_t igraph_i_community_fast_label_propagation(const igraph_t *graph,
+    igraph_vector_int_t *membership,
+    igraph_neimode_t mode,
+    const igraph_vector_t *weights,
+    igraph_vector_bool_t *fixed) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph), no_of_not_fixed_nodes = 0;
+    igraph_integer_t i, j, k;
+    igraph_inclist_t il;
+    igraph_adjlist_t al;
+
+    igraph_vector_t label_counters;
+    igraph_vector_int_t dominant_labels, nonzero_labels, node_order;
+    igraph_dqueue_t queue;
+    igraph_vector_bool_t in_queue;
+    igraph_neimode_t reverse_mode;
+
+    reverse_mode = IGRAPH_REVERSE_MODE(mode);
+
+    if (weights) {
+        IGRAPH_CHECK(igraph_inclist_init(graph, &il, reverse_mode, IGRAPH_LOOPS_ONCE));
+        IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+    } else {
+        IGRAPH_CHECK(igraph_adjlist_init(graph, &al, reverse_mode, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE));
+        IGRAPH_FINALLY(igraph_adjlist_destroy, &al);
+    }
+
+    /* Create storage space for counting distinct labels and dominant ones */
+    IGRAPH_VECTOR_INIT_FINALLY(&label_counters, no_of_nodes);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&dominant_labels, 0);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&nonzero_labels, 0);
+    IGRAPH_CHECK(igraph_vector_int_reserve(&dominant_labels, 2));
+
+    RNG_BEGIN();
+
+    /* Initialize node ordering vector with only the not fixed nodes */
+    IGRAPH_CHECK(igraph_dqueue_init(&queue, no_of_nodes));
+    IGRAPH_FINALLY(igraph_dqueue_destroy, &queue);
```

Comment:
Please use `IGRAPH_DQUEUE_INT_INIT_FINALLY` (after also changing the queue to store integers).